design/design-doc-terrajet.md

```diff
@@ -0,0 +1,526 @@
+logic for merge/diff operations. Though it doesn't support doing these operations
+for nested schema structs.
+
+### Interact with Terraform CLI
+
+We can produce a blob of JSON and give that to Terraform CLI for processing whenever
+we need to execute an operation. [As mentioned][json-hcl-compatibility] in
+Terraform docs, JSON can be a full replacement for HCL and thanks to [`jsoniter`][jsoniter]
+library, we'll be able to have multiple tag keys to use in marshall/unmarshall
+operations depending on the context. So, we will not need to generate any conversion
+functions to get JSON representation with snake case.
+
+Additionally, this option has the advantage that we'd be interacting with
+Terraform just like any user, hence all features, stability guarantees and
+community support would be available to us. For example, we won't need to build
+logic for importing resources or comparison tooling to see whether an update is
+due; we'll call the CLI and parse its JSON output.
+
+A disadvantage of this approach is that we'd need to work with the filesystem to
+manage JSON files and `.tfstate` file that contains the state. The footprint of
+multiple instances of Terraform CLI working at the same time could also bring in
+resource problems, however, we have workarounds for these problems like pooling
+the executions. Additionally, forking the CLI to make it work with an existing
+provider server is an option we can consider as well.
+
+## Proposal
+
+From the three options listed, interacting with the CLI is the one that will get
+us to the finish line fastest without too much compromise. Let's inspect the main
+drivers of this decision.
+
+#### Reimplementation of CLI
+
+Terraform CLI has generic functionality for operations it performs and providers
+only inject these operations with custom functions. For example, resources declare
+[custom diff][custom-diff-example] functions so that Terraform CLI is aware of
+resource-specific cases when it runs its `plan` logic. Another example would be
+[custom import][custom-import-example] logic. As you can see, there is only a small
+exception handling here so if we interact at this level we'd need to build `import`
+logic and then inject this. So, while it's intriguing to see CRUD operations
+exported right away, it's the other operations that need more effort and time.
+Interacting with CLI means we get to use Hashicorp's implementation, which is
+well-tested by the community, and not spend cycles.
+
+#### Unfriendly Programming Interfaces
+
+Since the providers are designed to be consumed only by Terraform CLI, they are
+designed for its specific needs and whatever the CLI doesn't need is not exposed.
+For example, [`ResourceDiff`][resource-diff] struct here has many fields and
+functions that are not exposed. So the implementation we'd be doing if we went
+with other two options will be very similar to how Terraform CLI implemented these
+functions already.
```

Comment:
It's also worth pointing out somewhere in this design that one of the principles in Crossplane's providers is a faithful representation of the external provider API.  

By adopting Terraform APIs, intermediary purpose-driven APIs are adopted, with the inclusion of behavioral fields that may (?) operate differently when executed in Crossplane.

Reply:
> By adopting Terraform APIs, intermediary purpose-driven APIs are adopted, with the inclusion of behavioral fields that may (?) operate differently when executed in Crossplane.

I think I follow what you're saying here, but could you give an example or two?

Reply:
The Crossplane interface to Terraform-derived providers will not be a faithful representation of the cloud provider API. Rather they will be faithful representations of the Terraform provider API.

The update of `ForceNew` behavior of a resource could be toggled by resource fields:
- https://registry.terraform.io/providers/hashicorp/aws/latest/docs/guides/resource-tagging#ignoring-changes-in-all-resources
- `instance_interruption_behavior` (as another example) https://github.com/hashicorp/terraform-provider-aws/blob/687cdcaa7a8c1b6f11e2f63431014533dd25aaec/aws/resource_aws_spot_instance_request.go#L140-L158

These fields are not part of the AWS API, they are helpers to achieve the desired Terraform behavior. How should this design doc address the shift from faithful API representation and how will this design be impacted or aided by these types of fields.

Reply:
Got it. My hunch is that we're effectively just going to 'leak' these Terraform details in our representation of cloud APIs. That's okay for the alpha stage, though eventually I suspect it will be a priority for the Terraform-derived MRs to match our 'native' MRs.

@muvaf perhaps we should consider a manually created filter of fields that don't map to the upstream APIs?

Reply:
We will have a way for provider authors to provide exception information here, like removal/addition of certain fields, manipulation of field tags/comments etc. So, this would be addressed by the provider authors, unfortunately, per resource if it's not generic enough. @displague thanks for the detailed explanation here, I'll add a section addressing this concern.

Reply:
Done!